internal/datastore/crdb/tx.go

```diff
@@ -77,32 +116,65 @@ func execute(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transac
 
 		// RELEASE SAVEPOINT itself can fail, in which case the entire
 		// transaction needs to be retried
-		if _, err := tx.Exec(ctx, "RELEASE SAVEPOINT cockroach_restart"); err != nil {
+		if _, err := tx.Exec(ctx, sqlReleaseSavepoint); err != nil {
 			return err
 		}
 		return nil
 	}
 
+	localTx := *tx
+	var i int
 	for i = 0; i < maxRetries; i++ {
-		if err = releasedFn(tx); err != nil {
-			if !retriable(ctx, err) {
-				return err
-			}
-			if _, retryErr := tx.Exec(ctx, "ROLLBACK TO SAVEPOINT cockroach_restart"); retryErr != nil {
-				return fmt.Errorf(errUnableToRetry, err)
+		if err = releasedFn(localTx); err != nil {
+			if retriable(ctx, err) {
+				if _, retryErr := localTx.Exec(ctx, sqlRollback); retryErr != nil {
+					return i, retryErr
+				}
+				continue
 			}
-			continue
+			return i, fmt.Errorf(errUnableToRetry, err)
 		}
-		return nil
+		return i, nil
 	}
-	return errors.New(errReachedMaxRetry)
+	return i, errors.New(errReachedMaxRetry)
+}
+
+// resetExecution attempts to rollback the given tx, begins a new tx with a new connection, and creates a savepoint.
+func resetExecution(ctx context.Context, conn conn, tx *pgx.Tx, txOptions pgx.TxOptions) (newTx pgx.Tx, err error) {
+	if *tx != nil {
```

Comment:
is `pgx.Tx` itself a pointer type? If so, why are we taking in a pointer to the pointer?